scene/2d/shape_cast_2d.cpp

```diff
@@ -0,0 +1,466 @@
+#endif
+		} break;
+		case NOTIFICATION_INTERNAL_PHYSICS_PROCESS: {
+			if (!enabled) {
+				break;
+			}
+			_update_shapecast_state();
+		} break;
+	}
+}
+
+void ShapeCast2D::_update_shapecast_state() {
+	result.clear();
+
+	ERR_FAIL_COND_MSG(shape.is_null(), "Invalid shape.");
+
+	Ref<World2D> w2d = get_world_2d();
+	ERR_FAIL_COND(w2d.is_null());
+
+	PhysicsDirectSpaceState2D *dss = PhysicsServer2D::get_singleton()->space_get_direct_state(w2d->get_space());
+	ERR_FAIL_COND(!dss);
+
+	PhysicsDirectSpaceState2D::ShapeParameters params;
+	params.shape_rid = shape_rid;
+	params.transform = get_global_transform();
+	params.motion = get_global_transform().basis_xform(target_position);
+	params.margin = margin;
+	params.exclude = exclude;
+	params.collision_mask = collision_mask;
+	params.collide_with_bodies = collide_with_bodies;
+	params.collide_with_areas = collide_with_areas;
+
+	if (target_position != Vector2()) {
+		dss->cast_motion(params, collision_safe_distance, collision_unsafe_distance);
+	}
+	bool intersected = true;
+	while (intersected && result.size() < max_results) {
+		PhysicsDirectSpaceState2D::ShapeRestInfo info;
+		intersected = dss->rest_info(params, &info);
+		if (intersected) {
+			result.push_back(info);
+			params.exclude.insert(info.rid);
+		}
+	}
+	collided = !result.is_empty();
+}
+
+void ShapeCast2D::force_shapecast_update() {
+	_update_shapecast_state();
+}
+
+void ShapeCast2D::add_exception_rid(const RID &p_rid) {
+	exclude.insert(p_rid);
+}
+
+void ShapeCast2D::add_exception(const Object *p_object) {
+	ERR_FAIL_NULL(p_object);
+	const CollisionObject2D *co = Object::cast_to<CollisionObject2D>(p_object);
+	if (!co) {
+		return;
+	}
+	add_exception_rid(co->get_rid());
+}
+
+void ShapeCast2D::remove_exception_rid(const RID &p_rid) {
+	exclude.erase(p_rid);
+}
+
+void ShapeCast2D::remove_exception(const Object *p_object) {
+	ERR_FAIL_NULL(p_object);
+	const CollisionObject2D *co = Object::cast_to<CollisionObject2D>(p_object);
+	if (!co) {
+		return;
+	}
+	remove_exception_rid(co->get_rid());
+}
+
+void ShapeCast2D::clear_exceptions() {
+	exclude.clear();
+}
+
+void ShapeCast2D::set_collide_with_areas(bool p_clip) {
+	collide_with_areas = p_clip;
+}
+
+bool ShapeCast2D::is_collide_with_areas_enabled() const {
+	return collide_with_areas;
+}
+
+void ShapeCast2D::set_collide_with_bodies(bool p_clip) {
+	collide_with_bodies = p_clip;
+}
+
+bool ShapeCast2D::is_collide_with_bodies_enabled() const {
+	return collide_with_bodies;
+}
+
+Array ShapeCast2D::_get_collision_result() const {
+	Array ret;
+
+	for (int i = 0; i < result.size(); ++i) {
+		const PhysicsDirectSpaceState2D::ShapeRestInfo &sri = result[i];
+
+		Dictionary col;
+		col["point"] = sri.point;
+		col["normal"] = sri.normal;
+		col["rid"] = sri.rid;
+		col["collider"] = ObjectDB::get_instance(sri.collider_id);
+		col["collider_id"] = sri.collider_id;
+		col["shape"] = sri.shape;
+		col["linear_velocity"] = sri.linear_velocity;
+		// col["metadata"] = sri.metadata;
```

Comment:
You can remove this line completely, metadata is not supported anymore.

Reply:
Is there a replacement/alternative to metadata in Godot 4.x? It's needed for determining which `TileMap` quadrants get affected by collision detection in Godot 3.x (see use case at /godotengine/godot-proposals#2543).

Reply:
Yes, it should be even a bit more clear since it's documented now.

In 4.0 each tile has a separate body, and you can access the corresponding tile coordinates based on the body id directly from the tilemap:
https://github.com/godotengine/godot/blob/284b50cf30a939dfee91876e06895aa50940c37c/doc/classes/TileMap.xml#L102-L108

Reply:
> Yes, it should be even a bit more clear since it's documented now.
> 
> In 4.0 each tile has a separate body, and you can access the corresponding tile coordinates based on the body id directly from the tilemap:
> 
> https://github.com/godotengine/godot/blob/284b50cf30a939dfee91876e06895aa50940c37c/doc/classes/TileMap.xml#L102-L108

Is there a way to get the tile's layer as well?
In order to retrieve custom data from the tile (from the custom data layer), we need to know the tile atlas coords ( aka because We need to call get_tile_data func on a TilesetAtlasSource, and then call get_custom_data on the TileData).

This is needed for any system that wants to identify the tile in a collision while also using layers.  Otherwise, all tiles will need to live in one layer or a known layer of the collision source